Automation

jamesscottbrown opened the pull request:
This PR should resolve #114 and #119.

When a **PR is created against master**, this:

* builds a PDF of the spec, and uploads a zip archive of this as a build artefact
* builds the samplers, and uploads a zip archive of these as a build artefact
* create separate zip files containing SVG/PNG/PDF files of each glyph, and upload these as build artefacts

(See examples from [a previous run](https://github.com/SynBioDex/SBOL-visual/actions/runs/390268742))

When a **push to master** happens (including when a PR is merged), it also:

* commits the `sampler/` directory (containing the automatically generated images). This means the website will be able to link to the image files in GitHub directly.
* updates `SEPs/sumamry.csv`, and commits the modified file


I think the build artifacts might only be available on GitHub for a limited time, so the PDF/PNG/SVG zip files should be copied to the SBOL website (they could be automatically copied to online storage e.g. by uploading them to an S3 bucket or `scp`ing them  to a server).

(I suggest closing this PR with **squash and merge**, as there are lots of trivial commits from trying to get this working that will clutter up the history.)

This also records the version history in the file listing the glyphs, so that the autogenerated `.tex` files in `specification/glyphscript` do not need manual modification to add macros like `\twotwozeronopage`.

jamesscottbrown committed:
Fix some SVG glyph files

* avoid setting atributes to 'None'
* fix offset in Aptamer glyph

jamesscottbrown committed:
initial versions of scripts

jamesscottbrown committed:
record versioning information and propogate automatically through to spec

jamesscottbrown committed:
delete commented-out code

jamesscottbrown committed:
add requirements.txt

jamesscottbrown committed:
move version amcro names from json file to script

jamesscottbrown committed:
generate sampler images in separate directory

jamesscottbrown committed:
update workflow file

jamesscottbrown committed:
separate out 2 workflow jobs

jamesscottbrown committed:
add a sudo command

jamesscottbrown committed:
fix YAML syntax

jamesscottbrown committed:
fix a path

jamesscottbrown committed:
pip -> pip3

jamesscottbrown committed:
add setuptools as a dependency

jamesscottbrown committed:
correct filename

jamesscottbrown committed:
commit sampler files on push/merge to master

jamesscottbrown committed:
add scripts/README.md

jamesscottbrown committed:
scripts/README.md

jamesscottbrown committed:
also create zipfiles containing all glyph images in a prticualr format

jamesscottbrown committed:
fix copy/paste error

jamesscottbrown committed:
delete the replaced glyscript shell scripts

jamesscottbrown committed:
add script to generate SEP table

jamesscottbrown committed:
update glyph_list.json to add new glyphs and record those that have changed

jakebeal commented:
Looking at the artifacts, I notice a couple of minor issues:

 - In the sampler, the "Functional Components" collection should be called "Sequence Features" and the "Species" collection should be called "Molecular Species"
 - Since the glyph zip collection is flat, glyphs with the same name (e.g., `no-glyph-assigned`) collide with one another. That's not necessarily a problem right now, since I think all such cases happen to be identical, but should we nest the collection or generate warnings or something?

jakebeal commented on scripts/glyph_list.json:
```diff
@@ -0,0 +1,79 @@
+[
```
Does this file need to be manually maintained, or is it being generated by one of the scripts?
If it's manual, it should be added to the SEP V023 draft.

jamesscottbrown replied:
It's currently manually maintained, but I'm not sure whether it should eventually be be replaced by the ontology (which would need to be automatically updated, and also expanded to record the version of the spec that last changed each glyph), or whether it should be used when generating the ontology.

jakebeal replied:
That's a question for discussion with @goksel that can be deferred until after this release; for now, can you please add a note to SEP V023 that will let us track this in workflow development?

jamesscottbrown replied:
I've added a comment.

jakebeal commented on scripts/rebuild_glyphs.py:
```diff
@@ -0,0 +1,57 @@
+script_path = os.path.realpath(__file__)
+script_dir = os.path.abspath(os.path.join(script_path, os.pardir))
+
+cmd_path = os.path.abspath(os.path.join(script_dir, os.pardir, "node_modules", "markdown-pdf", "bin", "markdown-pdf"))
```
Is PDF generation by this command idempotent? I ask because it's being run unconditionally on all glyphs, including ones that have not changed.  With the prior PDF conversion, that would generate diffs in every PDF every time, polluting commits.

jamesscottbrown replied:
I've just pushed a change that looks at the mtime for the PDF and the source directory (containing the README and image files), and skips the PDF generation if the PDF was generated more recently.

The PDFs generated by this script get used in the automatic building of the complete specification PDF, but aren't (currently) automatically committed to the repository.

jakebeal replied:
What is the workflow for a manual update, then? I want to make sure that it's still possible to build the specification PDFs locally by hand as well, and that we aren't only dependent on GitHub actions.

jamesscottbrown replied:
You should be able to run the `rebuild_glyphs.py` script as a drop-in replacement for the previous shell scripts. It's a python script with no python package dependencies (the only modules it imports are part of the standard library), and it relies on the same external program for the PDF generation (`markdown-pdf`). 

So:

```bash
cd scripts
npm install markdown-pdf
python3 rebuild_glyphs.py
```

should regenerate only the updated PDFs in `specifications/glyphscript/Glyphs/`, and update the `.tex` files in `specifications/glyphscript/`.

jakebeal replied:
Can we get this into the appropriate README file(s), so this documentation is easy to find?  After that, I'm ready to hit merge on the pull request.

jamesscottbrown committed:
tweak sampler titles

jamesscottbrown committed:
don't regenerate PDF unless source files have a more recent modification date

jamesscottbrown committed:
warn if two or mroe glyph files have the same name

jamesscottbrown commented:
> In the sampler, the "Functional Components" collection should be called "Sequence Features" and the "Species" collection should be called "Molecular Species"

I've modified `glyph_list.json` to update these names.

> Since the glyph zip collection is flat, glyphs with the same name (e.g., no-glyph-assigned) collide with one another. That's not necessarily a problem right now, since I think all such cases happen to be identical, but should we nest the collection or generate warnings or something?

I've added a warning in these cases.

jamesscottbrown committed:
update README for scripts
